Gallery Block: Add Media Library button to the upload new image area

jorgefilipecosta opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/8309
Supersedes: https://github.com/WordPress/gutenberg/pull/9682

This PR adds a media library button in the upload new image are of the gallery block.
Trying to follow the design 2 proposed in https://github.com/WordPress/gutenberg/pull/9682#issuecomment-425435709 by @kjellr and approved by @karmatosed.

A new functionality that allows the gallery to open in library/add frame instead of the edit frame was added to MediaUpload, so when the user pressed the media library button in the add zone the user goes directly to the add to library section when using the edit gallery button on the toolbar the user goes to the edit gallery section.

## Description
I added a gallery;
I added some images;
I selected the gallery;
I checked that a new media library button exists in the add to gallery zone.
If I click on it I go directly to a frame that allows the addition of images from the library to the gallery.

## How has this been tested?

## Screenshot
<img width="997" alt="screenshot 2018-11-27 at 15 35 14" src="https://user-images.githubusercontent.com/11271197/49093136-64176280-f25b-11e8-82de-ff44cee4ab47.png">

jorgefilipecosta committed:
Gallery Block: Add Media Library buttons to the upload new image area; Media Upload: Allow galleries to open in the library frame.

simison commented:
Could this section be its own independent component so that other blocks could re-use it?

It would be super useful when building e.g. alternative gallery blocks or slider blocks.

talldan commented on packages/edit-post/src/hooks/components/media-upload/index.js:
```diff
@@ -78,6 +78,7 @@ class MediaUpload extends Component {
 		allowedTypes,
 		multiple = false,
 		gallery = false,
+		galleryOpenInLibrary = false,
```
I'm not convinced the prop name conveys the right meaning. I think renaming it to something like 'addToGallery' could be clearer.

talldan reviewed (commented):
Hey @jorgefilipecosta - this looks pretty close. Not sure if it's still WIP. I gave it a review anyway 😄 

Would be cool to get a designer to have a look over it. I was thinking a bit on how the button styles were different to the other buttons and wondering if they could be brought closer to or based off of other existing styles instead of just the default.

jasmussen commented:
I would love for us to revisit, holistically, the appender overall, to give more options and a better overall design going into phase 2. However that is probably best done separately, and not as part of this bugfix. 👍 👍for this as a good iteration.

jorgefilipecosta commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -284,17 +285,39 @@ class GalleryEdit extends Component {
+								gallery
+								value={ images.map( ( img ) => img.id ) }
+								render={ ( { open } ) => (
+									<IconButton
```
Hi @talldan, yes we have bugs in the media gallery that make it easily out of sync, I expect https://github.com/WordPress/gutenberg/pull/12435 to address these sync issues.

jorgefilipecosta commented on packages/edit-post/src/hooks/components/media-upload/index.js:
```diff
@@ -90,7 +91,12 @@ class MediaUpload extends Component {
-			const currentState = value ? 'gallery-edit' : 'gallery';
+			let currentState;
+			if ( galleryOpenInLibrary ) {
+				currentState = 'gallery-library';
```
Hi @talldan I did not find any documentation regarding this :( I checked what states we have by exploring the core code.

jorgefilipecosta commented:
Hi,
I'm sorry for the delay in answering I missed the notifications I think I probably unsubscribed to them by mistake.

> Could this section be its own independent component so that other blocks could re-use it?

> It would be super useful when building e.g. alternative gallery blocks or slider blocks.

> The component could even include the `drop-zone` component and support props supported by [`media-placeholder`](https://github.com/WordPress/gutenberg/tree/fe51bd08684462690988de7ec61a3cdda083710a/packages/editor/src/components/media-placeholder) component. It would basically be generic "add additional media to this block" -wrapper.

I agree, my plan is, for now, merge the enhancements in the gallery get them a little bit tested and then after everything is polished extract and expose a generic component.

Hi @talldan, thank you for your review, I think I addressed all your comments.

kjellr commented:
Hey @jorgefilipecosta, thanks for reviving this PR. This works well in my testing!

A couple small design notes: 

<img width="682" alt="screen shot 2019-02-18 at 8 57 14 am" src="https://user-images.githubusercontent.com/1202812/52956036-31fa3600-335c-11e9-8a87-77295b01eb7d.png">

1. The button styling is a little odd — the text is blue, and it almost looks disabled to me? I'd expect the button to share the color/border/shadow of our other secondary buttons: 

<img width="95" alt="screen shot 2019-02-18 at 8 56 46 am" src="https://user-images.githubusercontent.com/1202812/52955789-9c5ea680-335b-11e9-84c4-ae13eec9cda8.png">

2. There's a little extra margin below the dotted outline. It looks like it's inheriting the 16px bottom margin for gallery items, but that can be zeroed out so there's even spacing all around it.

jorgefilipecosta commented:
Thank you for the review @kjellr I will look into fixing these issues.

My plan was to after merging this PR extract a component that allows rendering the gallery appender. I noticed that the gallery appender being implemented here is just equivalent to `MediaPlaceholder` without label and icon, with the media library button opening the media library in a different state, and with some custom styles.
It may be possible refactor this changes to use `MediaPlaceholder` in the gallery appender by adding two additional optional props one that enables a custom style and one that allows opening the media library in a specific state (the add to gallery mode). I will check if this is, in fact, a possibility and update the PR accordingly.

simison commented:
Would it make sense if `MediaPlaceholder` could take a child component?

Or perhaps it should be called `MediaContainer` instead:

```jsx
<MediaContainer>
  <Gallery/>
</MediaContainer>
```

Where `<Gallery>` could be also `<Slideshow>` or `<TiledGallery>` for Jetpack or even something like `<VideoPlaylist>`. Child component would only care about `files` array and how to visualize it — the wrapper would take care of handling the files, dropzone, showing a place holder when there are no files. When there are files, wrapped would show the placeholder under the gallery component.

Something like (very much simplified):
```jsx
function MediaContainer(WrappedComponent) {
  return class extends Component {
    // handle state, file uploads, notices etc

    render() {
      if ( ! this.state.files.length ) {
        return <MediaPlaceholder />
      }

      return (
        <Fragment>
          { this.props.noticeUI }
          <DropZone />
          <WrappedComponent files={ this.state.files } { ...this.props } />
          <MediaPlaceholder title={ false } />
        </Fragment>
      );
    }
  };
}

export default withNotices( MediaContainer );
```

jorgefilipecosta commented:
The code was updated, now instead of custom add item logic MediaPlaceholder handles everything (it got new additions). The gallery code and styles got simplified with this change.
cc: @sirreal, @simison 

@kjellr The design was changed and I think your points were addressed:
<img width="1364" alt="screenshot 2019-02-20 at 17 03 46" src="https://user-images.githubusercontent.com/11271197/53112237-b6081580-3536-11e9-8f71-354311900885.png">

Hi @simison,
Right now with this change the `MediaPlaceholder` already fits most use cases and some functions like the one that handles the addition of new files are not required anymore.
In the idea, you are proposing `MediaContainer` is responsible for rendering the MediaPlaceholder, the MediaPlaceholder contains many block specific parts (icon, labels, mapping of file/media objects to attributes, etc... This props would also need to be passed to `MediaContainer` so I'm not sure if the gain is significant.
I think we can try to enhance `MediaPlaceholder` and then we may think of ways to generalize more.

sirreal commented on packages/editor/src/components/media-placeholder/index.js:
```diff
@@ -137,26 +152,25 @@ export class MediaPlaceholder extends Component {
 		}
 
-		if ( ! instructions || ! title ) {
+		if ( instructions === undefined || title === undefined ) {
```
This is likely to break backwards compatibility, isn't it? Folks that have been relying on the default labels will no longer get them.

An alternative would be to accept `null` for these props so that folks can explicitly opt-out of the default label and instructions. It's much less likely folks are using a falsey value other than `undefined` to get the defaults.

jorgefilipecosta replied:
> This is likely to break backward compatibility, isn't it? Folks that have been relying on the default labels will no longer get them.

I think folks that relied on the defaults will still get them because in that case instructions or title are undefined.

What this allows is the ability to not render instructions or title by setting a null value or empty string in these fields.
There is minor behavior change previously if title/instructions were null or contained an empty string we applied default value now for this cases we don't. But I don't think someone would set the prop to a null/empty string and expect its value to contain the defaults.

sirreal commented on packages/editor/src/components/media-placeholder/index.js:
```diff
@@ -193,43 +207,74 @@ export class MediaPlaceholder extends Component {
 						{ __( 'Upload' ) }
 					</FormFileUpload>
 					<MediaUpload
+						addToGallery={ addToGallery }
```
Is there a use case when we'd disable `addToGallery`? I'd expect folks to always need to opt-in to this behavior when using the `MediaPlaceholder`.

It also seems to overlap with `isAppender`, although one is for styling and the other behavior.

jorgefilipecosta replied:
> Is there a use case when we'd disable `addToGallery`? I'd expect folks to always need to opt-in to this behavior when using the `MediaPlaceholder`.

When the gallery/value was empty previously we went to the create gallery flow instead of the add to gallery one. If we default to addToGallery we would be applying a behavior change and go by default to the add flow instead of the create flow.

> It also seems to overlap with `isAppender`, although one is for styling and the other behavior.

Yes, I preferred to have a flag for behavior and another one for the style. So we allow more customization.

Thank you for the review @sirreal 👍

sirreal reviewed (commented):
I really like the recent changes, the `MediaPlaceholder` component becomes much more useful 🙌

kjellr commented:
Thank you, @jorgefilipecosta! 

I pushed a small update to the border styles, to sync them up with the final styles used in #10136: 

![screen shot 2019-02-21 at 3 24 22 pm](https://user-images.githubusercontent.com/1202812/53199379-f1651b80-35ec-11e9-8f06-5f0c64002b93.png)

Other than that, just a couple small bugs: 

1. It looks like there's [a bit of extra spacing](https://cloudup.com/cunymfAiRw6) on the bottom of the hover state.
2. I think we've lost the Media + Text icon from the button in this iteration. We can keep that in. 

---

Also, I had a functional question after looking back in on #10136. It looks like the entire appender from that ticket is clickable. Should this one behave the same way? Right now, nothing happens unless you click on the "Upload" or "Media Library" buttons: 

![focus-2](https://user-images.githubusercontent.com/1202812/53199830-173ef000-35ee-11e9-9f02-4880b167b9d7.gif)

I think it might be nice to let people click anywhere on the appender (except for the Media Library button) to open up the file dialog.

cc @jasmussen since you may have some thoughts here too.

jasmussen commented:
Phew, tricky one, this. 

I think we might need to find a simple interim solution and get this improvement shipped, and then look at a long term design improvement for this appender separately. The following is based on the great thoughts by Kjell. How about we do this:

- The dashed area is the drag and drop dropzone. It highlights when you drop an image on it. You can also click it to open the media library, this is a standard pattern.
- Or you could click the Media Library button separately.
- Or you could click the Upload button separately.
- Both buttons are standard gray buttons. And either both buttons have an icon, or none of them do. 

Does that sound shippable?

simison commented:
> It highlights when you drop an image on it. You can also click it to open the media library, this is a standard pattern.

Wouldn't it be more logical to have _upload_ modal (and not the media gallery) open when clicking the area where you can upload files by dropping? Suppose that's why "upload" button didn't have background, to highlight the fact that it has the same feature as the whole dropzone area?

jasmussen commented:
> Wouldn't it be more logical to have _upload_ modal (and not the media gallery) open when clicking the area where you can upload files by dropping?

Sure that makes sense. 

But I don't think the background should highlight the button, they are two separate things, and the separate button is good to provide regardless. I also imagine it would visually simplify by having two buttons. 

The important thing isn't so much the intricacies, but just picking a style and moving, because we know we want to revisit this in a more holistic way.

kjellr commented:
Thanks as always for the clarifying thoughts, @jasmussen. That all makes sense, and I like @simison's suggestion to have the click trigger the upload modal instead of the Gallery one. 

---

> The dashed area is the drag and drop dropzone. It highlights when you drop an image on it. You can also click it to open the media library, this is a standard pattern.

This makes sense. One thought on the drop zone though: Currently, dropping an image _anywhere_ on the gallery block is fine: 

![drop](https://user-images.githubusercontent.com/1202812/53254757-06e05100-3692-11e9-96ed-2d645cb4f739.gif)

Do we want to limit that to just the drop zone area? 

![drop-2](https://user-images.githubusercontent.com/1202812/53254768-0d6ec880-3692-11e9-8a42-02e0e506b80e.gif)

---

> Both buttons are standard gray buttons. And either both buttons have an icon, or none of them do.

I just pushed a small commit to restore the standard gray style to the upload button. 👍 

Regarding the icons though, after clicking around to what we're doing elsewhere, I think it's fine to leave off the Media Library icon. All of our other placeholders have an icon for upload, and no icon for the other buttons: 

<img width="640" alt="screen shot 2019-02-22 at 10 59 44 am" src="https://user-images.githubusercontent.com/1202812/53254241-faa7c400-3690-11e9-9a1e-a4f7b0dbb2f7.png">

(We could change this of course, but we'd probably want to change it everywhere.)

jorgefilipecosta commented:
Thank you for the fixes @kjellr, I applied a fix to hover space.

> I think we've lost the Media + Text icon from the button in this iteration. We can keep that in.

> Regarding the icons though, after clicking around to what we're doing elsewhere, I think it's fine to leave off the Media Library icon

So, for now, we can leave the icon right?

kjellr commented:
> So, for now, we can leave the icon right?

Yeah, I think that's fine for now. It's consistent with the way these buttons appear elsewhere.

simison commented:
BTW just noting that core gallery block in `master` has no more block-wide button, but a kind of "image placeholder" button instead:

<img width="675" alt="image" src="https://user-images.githubusercontent.com/87168/53573527-546b2b00-3b76-11e9-9659-bf390cf3e2be.png">

...does the design here need to take this into account?

simison commented:
This now needs rebase after https://github.com/WordPress/gutenberg/pull/14115

aduth commented on packages/editor/src/components/media-upload/README.md:
```diff
@@ -101,6 +101,17 @@ CSS class added to the media modal frame.
+
+If true the gallery media modal will open in directly in the media library where additional images can be added.
+If false the gallery media modal will open in the edit mode where existing images can be edited.
+Only applies if gallery = true.
```
Leaving this comment very much in passing, without looking much at context: The interface of this component doesn't otherwise reference "gallery" terminology, but rather "multiple". There is no "gallery = true"

jorgefilipecosta replied:
Hi @aduth, gallery and multiple are not the same thing. Multiple allows the selection of multiple files using the normal media modal frames, while the gallery prop opens specific gallery frames optimized for the gallery usage.
The gallery prop is not something we are adding here it is something that already existed before:
https://github.com/WordPress/gutenberg/blob/9a93e6a49fa86805e9e0a8bed6f102787e4d2e36/packages/block-library/src/gallery/edit.js#L200-L213


But, the documentation of the interface missed the gallery prop so it was misleading, I added the documentation to this prop.

jorgefilipecosta commented:
Hi @simison,
I remember that gallery had a design for the add button equivalent to the one you shared but it was meanwhile changed and the add button is already wide on master:
<img width="1111" alt="Screenshot 2019-03-15 at 08 50 22" src="https://user-images.githubusercontent.com/11271197/54419795-38cc5c80-4700-11e9-9d92-edada135fd39.png">


I rebased the PR did a new set of test. I reverted the fix to hoover spacing problem -- the fix caused a bug when the gallery is in the placeholder state, and I noticed this problem is already present on master. I would prefer if we address the problem separately.

@jasmussen @kjellr any thoughts regarding the current state of this PR? Any task remaining in your opinion?

kjellr commented:
> @jasmussen @kjellr any thoughts regarding the current state of this PR? Any task remaining in your opinion?

This is looking good. Just one remaining item in my opinion: We should set it so that a click anywhere on the drop zone triggers the upload modal. 

I do feel like we should consider changing the dropzone area as described above https://github.com/WordPress/gutenberg/pull/12367#issuecomment-466450977, but that'd come with some tradeoffs, so I think it's fine to punt it for now and revisit later. 👍

jorgefilipecosta committed:
Allow clicking all placeholder appender area to upload files.

jorgefilipecosta commented:
Hi @kjellr, this PR was updated and all the appender area is clickable.

kjellr committed:
Switch to a "pointer" cursor on the appender on hover.

To reinforce that this area is clickable.

kjellr committed:
Reinstate the gray background.

kjellr reviewed (approved):
@jorgefilipecosta This is looking great! I pushed two small changes: 

- I added `cursor: pointer` to the hover state, to reinforce that the whole area is now clickable. 
- I reinstated the gray background to sync up with the other inserters we're adding in 
https://github.com/WordPress/gutenberg/pull/14241.
- _EDIT:_ I also added 1fdb53f, which switches the hover state to use a `$white` outline color on dark themes, to align with #14241. 

![Screen Shot 2019-03-25 at 8 41 28 AM](https://user-images.githubusercontent.com/1202812/54920749-67092380-4eda-11e9-87a6-fefb484355fc.png)

This should be ready to go!

kjellr committed:
Switch to a white dashed outline on dark backgrounds.

jorgefilipecosta committed:
Updated the changelogs.
